script_path <- this.path::this.path()
script_dir <- dirname(script_path)

util <- paste0(script_dir,'/../../ut/util.r')
treetool <- paste0(script_dir,'/../tree/treetools.r')

suppressWarnings({
    source(util)
    source(treetool)
    library(argparse)
})

parser <- ArgumentParser()
parser$add_argument("--multigene_table", help="Path to multigene table")
parser$add_argument("--tree_dir", help="Path to tree directory")
parser$add_argument("--metadata", help="Path to metadata")
parser$add_argument("--uhgg_metadata", help="Path to UHGG metadata")
parser$add_argument("--sf_dir", help="Path to StrainFinder directory")
parser$add_argument("--phylo_enrichment_dir", help="Path to directory containing phylo enrichment results")
parser$add_argument("--output_dir", help="Path to output directory")
parser$add_argument("--genomes", help="List of genomes that we run StrainFinder on")
parser$add_argument("--strainfinder_tips", help="StrainFinder tips used")
args <- parser$parse_args()

genomes.use = args$genomes
sf.tips = args$strainfinder_tips
sf.dir = args$sf_dir 

#########################################
# GENERATE LIST OF NON REDUNANT GENOMES #
#########################################

# ----------------------------------
# run 1.select_strains.r to generate
#  1. 'genomes.use.txt', and 
#  2. 'strainfinder.final_tips.txt'
# ----------------------------------

# load genomes
g = readLines(genomes.use)

# load strain stats
y = ffread(sf.tips, as.dt=T)
setkey(y, name)

# ---------------------------------------------------
# PARSE StrainFinder RESULTS
# 
# generated by calling: 2.merge_np.py and 3.run_sf.py
# ---------------------------------------------------

# load strain abundances
x = sapply(g, function(gi){
    if(file.exists(paste0(sf.dir, gi, 'sf.txt.3'))){
	xi=read.table(paste0(sf.dir, gi, 'sf.txt.3'), sep='\t', header=F, row.names=1)
	ni=y[name == gi]$node
	rownames(xi) = gsub('_[^_]*$', '', rownames(xi))
	colnames(xi) = c(sort(as.character(ni)), 'other')
	xi
    }
}, simplify=F)

# load metadata
meta = read.metadata.nice(args$metadata)
rownames(meta) = meta$sample

# load bacterial metadata
bact = ffread(args$multigene_table, as.dt=T)
setkey(bact, genome)

#----------------------------------------------------------------
# build dataframe of useful stats
#  * output: 'multigene.nr.txt' this is very importnant for plots
#----------------------------------------------------------------

# for every genome...
g.use = names(which(!sapply(x, is.null)))
res = sapply(g.use, function(gi){
    
    node = y[name == gi]$node
    freq = x[[gi]]
    
    samp = unname(sapply(y[name == gi]$samp, function(a) gsub('_[^_]*$', '', strsplit(a, ';')[[1]])))
    names(samp) = as.character(node)
    di = meta[rownames(freq),'nice_dx']
    
    # abundances within vs. between samples
    u = data.frame(t(sapply(colnames(freq), function(a){
        
        # abundance inside/outside node
	if(a != 'other'){
            ab.in = mean(freq[samp[[a]],a])
	    ab.out = mean(freq[setdiff(rownames(freq), samp[[a]]),a])
	} else {
	    ab.in = NA
	    ab.out = NA
	}
        
	# abundance health/disease
	i = di == 'HC'
	ab.h = mean(freq[i,a])
	ab.d = mean(freq[!i,a])
        
	# abundance high/low
	ab.lo = mean(freq[,a][freq[,a] <= .5])
	ab.hi = mean(freq[,a][freq[,a] >= .5])
	ab.5 = mean(freq[,a] <= .05)
	ab.25 = mean(freq[,a] <= .25)
	ab.50 = mean(freq[,a] <= .50)
	ab.95 = mean(freq[,a] <= .95)
        
	# expression stats
	ab.mean = mean(freq[,a], na.rm=T)
	ab.prev = mean(freq[,a] > .5, na.rm=T)
	
	list(ab.in=ab.in, ab.out=ab.out, ab.h=ab.h, ab.d=ab.d, ab.lo=ab.lo, ab.hi=ab.hi, ab.5=ab.5, ab.25=ab.25, ab.50=ab.50, ab.95=ab.95, ab.mean=ab.mean, ab.prev=ab.prev)
    })))
    
    cbind(name=gi, node=colnames(freq), u)
    
}, simplify=F)

# fix results
res = as.data.frame(do.call(rbind, res))
res[,3:ncol(res)] = sapply(res[,3:ncol(res)], function(a) as.numeric(unlist(a)))
res = as.data.table(res)

# merge with statistics
y$node = as.character(y$node)
res = merge(res, y, by=c('name', 'node'), all=T)
res$id = paste(res$name, res$node, sep='.')
res$group = ifelse(res$node == 'other', 'other', ifelse(!is.na(res$padj), 'sig', 'non'))
res = setkey(res, id)

# merge with annotations
colnames(bact)[[1]] = 'name'
res = merge(res, bact, by='name', all=T)

# fix redundant genomes
nr = unique(res[,.(name, dnaG, gyrB, rpoB, u_dnaG, u_gyrB, u_rpoB, rho, padj)])[order(padj, -rho)]
nr = nr[,dup_dnaG := duplicated(dnaG)][,dup_gyrB := duplicated(gyrB)][,dup_rpoB := duplicated(rpoB)][,dup := (dup_dnaG | (dup_gyrB | dup_rpoB))]
writeLines(nr[dup == FALSE]$name, paste0(args$output_dir, '/multigene.nr.txt'))

# make non-redundant set of strains
# ---------------------------------
nrg = nr[dup == FALSE]$name
res = res[name %in% nrg][order(padj)]       

                           
#####################################
# MERGE WITH OTHER NODE LEVEL STATS #
#####################################

# helper function
simple_select_nodes = function(tree, nodes, min_size=25, other=TRUE){

    # get disjoint significant nodes
    nodes.use = c()
    for(ni in nodes){
        flag = 1
        li = extract.clade(tree, ni)$tip.label
        for(nj in nodes.use){
            lj = extract.clade(tree, nj)$tip.label
            if(length(intersect(li, lj)) > 0){
                flag = 0
                break
            }
        }
        if(flag == 1){
            nodes.use = c(nodes.use, ni)
        }
    }

    if(other == FALSE){
        return(nodes.use)
    }

    # get disjoint other nodes
    n0 = length(tree$tip.label)+1
    nf = n0 + tree$Nnode - 1
    node2size = sort(setNames(sapply(n0:nf, function(a) Ntip(extract.clade(tree, a))), n0:nf), dec=T)
    for(ni in names(node2size)){ni = as.integer(ni)
        if(node2size[[as.character(ni)]] >= min_size){
            flag = 1
            li = extract.clade(tree, ni)$tip.label
            for(nj in nodes.use){
                lj = extract.clade(tree, nj)$tip.label
                if(length(intersect(li, lj)) > 0){
                    flag = 0
                    break
                }
            }
            if(flag == 1){
                nodes.use = c(nodes.use, ni)
            }
        }
    }
    return(nodes.use)
}

# -----------------
# bacterial genomes
# -----------------
bact = ffread(args$multigene_table)

# fix taxonomic names
bact[, c('domain','phylum','class','order', 'family','genus')] = sapply(bact[, c('domain','phylum','class','order', 'family','genus')], function(a) gsub('\\_[A-Z]$','',a))
bact$family = gsub('Acutalibacteraceae','Ruminococcaceae', bact$family)

# fix non-redundant column
nr = readLines(paste0(args$output_dir, '/multigene.nr.txt'))
bact$nr = ifelse(bact$genome %in% nr, TRUE, FALSE)

# -----------------
# strain enrichment
# -----------------
tree_dir <- args$tree_dir
# enrichment stats
ph = do.call(rbind, sapply(bact$genome, function(gi) fread(sprintf('%s/%s.phylo.enrich.full.txt', args$phylo_enrichment_dir, gi)) ,simplify=F))
ph = ph[test == 'nice_dxIBD' | is.na(test)]
ph[, padj:=p.adjust(pval, 'fdr'), .(name)]
ph = ph[order(pval)]
ph = ph[,{i = padj < .05; i[which.min(padj)] = TRUE; .SD[i]},name]
ph = ph[order(padj)][,.(name, node, coef, padj)]
colnames(ph) = c('genome', 'node', 'coef', 'padj')

# phylogeny stats
u = sapply(1:nrow(ph), function(i){

  gi = ph[i,]$genome
  ni = ph[i,]$node
  x = read.tree.nice(sprintf('%s/%s.tree',tree_dir, gi))
  nt = Ntip(x)
  
  da = table(factor(gsub(".*_", "", x$tip.label), levels = c("HC", "CD", "UC")))
  da = c(Ha = unname(da[1]), Da=sum(da[2:3]))
  
  tips = extract.clade(x, ni)$tip.label
  db = table(factor(gsub(".*_", "", tips), levels = c("HC", "CD", "UC")))
  db = c(Hb = unname(db[1]), Db=sum(db[2:3]))
  
  da = da -db
  
  nb = length(tips)
  
  ifn = sprintf('%s/%s.ref.tree',tree_dir, gi)
  if(file.exists(ifn)){
    y = read.tree.nice(ifn)
    nj = node.map(ni,x,y)
    refs = extract.clade(y, nj)$tip.label
    refs = sum(grepl('^GUT', refs))
  } else {
    refs = 0
  }
  c(genome=gi, node=ni, tips = nt, da, db, n.below = nb, refs=refs)
  }, simplify=F)
u = data.frame(do.call(rbind,u))
u[,-1] = sapply(u[,-1], as.numeric)
u = data.table(u)

# merge results
# -------------
res = as.data.table(merge(bact, merge(ph, u, by=c('genome', 'node'), all=T), by='genome', all=T))
                           
# make non-redundant set of strains
res = res[order(padj)]
g.use = unique(res$genome)

ref = sapply((1:length(g.use)), function(i){gi = g.use[[i]]; print(i);

    # get non-redundant strains                                 
    ot = read.tree.nice(sprintf('%s/%s.tree',tree_dir, gi))
    ns = res[padj < .05][genome == gi]$node
    nd = sort(unique(simple_select_nodes(ot, ns, min_size=25)))
    
    # get reference genomes
    fn = sprintf('%s/%s.ref.tree',tree_dir, gi)
    if(file.exists(fn)){
        nt = read.tree.nice(fn)
        nj = setNames(node.map(nd, ot, nt), nd)
        ri = tryCatch({
            rf = sapply(names(nj), function(a){node = nj[[a]]
                ret = fix_ref_names(grep('^GUT', extract.clade(nt, node)$tip.label,value=T))
	        if(length(ret) == 0){ret = NA}
	        return(ret)
            }, simplify=F)
            rf = stack(rf)[,c('ind', 'values')]
            rf = cbind(gi, rf)
            colnames(rf) = c('genome', 'node', 'ref')
            rf$sig = rf$node %in% ns
	    rf
	}, error=function(e){
	    NULL
	})
    } else {
        ri = NULL
    }
    return(ri)
}, simplify=F)

ref = as.data.table(do.call(rbind, ref))
ref$node = as.integer(as.character(ref$node))
res = merge(res, ref, by=c('genome','node'), all=T)
ofn = paste0(args$output_dir, '/res.rds')
saveRDS(res, file=ofn)
                           
########################################
# ITERATE OVER ALL TREES AND SUMMARISE #
########################################

bact = ffread(args$multigene_table, as.dt=F)

meta = ffread(args$uhgg_metadata)
rownames(meta) = meta$Genome

# genomes with refs
gis  = bact$genome
ifn = sprintf('%s/%s.ref.tree',tree_dir, gis)
i = file.exists(ifn)
gis = gis[i]
ifn = ifn[i]

# signficiance of refs
sig = sapply(gis, function(gi) fread(sprintf('%s/%s.phylo.enrich.full.txt', args$phylo_enrichment_dir, gi)),simplify=F)
sig = do.call(rbind,sig)
             
# select genome, adjust pvalues
sig = sig[is.na(test) | test == 'nice_dxIBD']
sig[, padj:= p.adjust(pval,'fdr'), by=name]
sig = sig[order(name, padj,pval)]
             
# subset to signif genomes
gis = sig[, .SD[which.min(padj)], by=name][padj < 0.05]$name
             
df = do.call(rbind, sapply(gis, function(gi){
        ifn1 = sprintf('%s/%s.tree',tree_dir, gi)
        ifn2 = sprintf('%s/%s.ref.tree',tree_dir, gi)
    
        tree.A = read.tree.nice(ifn1)
        tree.B = read.tree.nice(ifn2)

        dx = sig[name==gi]
        old.nodes = dx$node
        new.nodes = node.map(old.nodes, tree.A, tree.B)
        ref.all = sum(grepl('^GUT', tree.B$tip.label))
        str.all = Ntip(tree.B) - ref.all

        tips = sapply(phangorn::Descendants(tree.B, new.nodes), function(vi) tree.B$tip.label[vi[vi<=Ntip(tree.B)]], simplify=F)

        ref.below = sapply(tips, function(a) sum(grepl('^GUT',a)))
        ref.above = ref.all - ref.below

        str.below = sapply(tips, length) - ref.below
        str.above = str.all - str.below
        ri = cbind(genome = gi, old.nodes=old.nodes, new.nodes=new.nodes, 
                   ref.below=ref.below, ref.above=ref.above, 
                   str.below=str.below, str.above=str.above, padj = dx$padj, coef = dx$coef)
        return(ri)
        }, simplify=F))
df = data.frame(df)
# fix numerics
df[,-1] = sapply(df[,-1], as.numeric)
ofn.tab = paste0(args$output_dir, '/stats.summary.txt')
writeTable(df, ofn.tab)

###############################
# TABULATE STRAINFINDER FREQS #
###############################

# load genomes
g = readLines(genomes.use)

# load metadata
meta = read.metadata.nice(args$metadata)
rownames(meta) = meta$sample
samples = rownames(meta)

# load strain stats
y = ffread(sf.tips, as.dt=T)
setkey(y, name)

# load strain abundances
x = sapply(g, function(gi){
    
    if(file.exists(paste0(sf.dir, gi, 'sf.txt.3'))){
        xi=read.table(paste0(sf.dir, gi, 'sf.txt.3'), sep='\t', header=F, row.names=1)
        ni=y[name == gi]$node
        rownames(xi) = gsub('_[^_]*$', '', rownames(xi))
        colnames(xi) = c(sort(as.character(ni)), 'other')
        xi
    }
}, simplify=F)

g.use = names(which(!sapply(x, is.null)))
x = sapply(g.use, function(gi){
        xi = gtools::na.replace(setRownames(x[[gi]][samples,],samples),0)
        xi},simplify=F)
x = do.call(cbind, x)
fwrite(x, paste0(args$output_dir,'/strain.freq.csv'), row.names=TRUE)
