script_path <- this.path::this.path()
script_dir <- dirname(script_path)

# SELECT IDS FOR MULTIGENE PANEL
util <- file.path(script_dir, "/../../ut/util.r")
treetools <- file.path(script_dir, "/../tree/treetools.r")
source(util)
source(treetools)

args = commandArgs(trailingOnly=TRUE)

# gi is the genome ID
gi = args[1]


#---------------------------------#
# please the modify the filepaths #
#---------------------------------#

# ifn1 is the nexus file generated by the multigene step
# ifn2 is the concatenated seq


ifn1 = sprintf('/broad/smillie-data/proj/mgxevo/amphora/phylogeny/multi/%s.conaln.nex', gi)
ifn2 = sprintf('/broad/smillie-data/proj/mgxevo/amphora/phylogeny/multi/%s.conaln.fas', gi)


ofn1 = sprintf('/broad/smillie-data/proj/mgxevo/amphora/phylogeny/filtered/%s.conaln.filt.nex', gi)
ofn2 = sprintf('/broad/smillie-data/proj/mgxevo/amphora/phylogeny/filtered/%s.conaln.filt.fas', gi)

mp = grep('charset',readLines(ifn1),value=T)
mp = gsub('charset','',mp)
mp = gsub('\\s|;','',mp)
mp = data.table(str_split_fixed(mp,'=',2))
mp[,c("beg","end"):=tstrsplit(V2, '-')]
mp[,gene:=V1]
mp[,V2:=NULL]
mp[,V1:=NULL]
mp[,beg:=as.integer(beg)]
mp[,end:=as.integer(end)]

ui = read.dna(ifn2, format = "fasta",as.matrix=TRUE,as.character=TRUE)
vi = Matrix(ui!='n', sparse=T)
pt = rep("NA", ncol(ui))
for(i in 1:nrow(mp)){
    pt[(mp[i]$beg):(mp[i]$end)] = mp[i]$gene
}

h  = grepl('HC',rownames(vi))
d  = grepl('IBDU|UC|CD',rownames(vi))
fh = colMeans(vi[h,])
fd = colMeans(vi[d,])

# sites that are roughly equal prop in IBD/HC
j = abs(log(fh/fd))<.5
j[is.na(j)] = FALSE
vi = vi[,j]
ui = ui[,j]
pt = pt[j]

# only samples that cover at least 20% of the multigene alignment
i = which.names(rowMeans(vi)>.2)
vi = vi[i,]
ui = ui[i,]

# write nex
nex = c('#nexus','begin sets;', sapply(unique(pt), function(pi){paste0('    charset ',pi,' = ',min(which(pt==pi)),'-',max(which(pt==pi)),';')}),'end;','')
fas = sapply(1:nrow(ui), function(i) paste0('>',rownames(ui)[i],'\n',paste0(ui[i,],collapse='')))
writeLines(nex, ofn1)
writeLines(fas, ofn2)            